File: java/Array/zeroestoend.java
import java.util.*;

class zeroestoend
{
    public static int[] movezeroes(int[] arr, int n)
    {
        // ArrayList<Integer> temp = new ArrayList<>();
        // for(int i=0;i<n;i++)
        // {
        //     if(arr[i]!=0)
        //         temp.add(arr[i]);
        // }
        // while(temp.size()<n)
        // {
        //     temp.add(0);
        // }
        // for(int i=0;i<n;i++)
        // {
        //     arr[i]=temp.get(i);
        // }
        // return arr;

        int j=0;
        for(int i=0;i<n;i++)
        {
            if(arr[i]!=0)
            {
                arr[j++]=arr[i];
            }
        }
        while(j<n)
        {
            arr[j++]=0;
        }
        return arr;
    }

    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] arr = new int[n];
        for(int i=0;i<n;i++)
        {
            arr[i]=sc.nextInt();
        }
        int[] ans=movezeroes(arr,n);
        for(int i=0;i<n;i++)
        {
            System.out.print(ans[i]+" ");
        }
        System.out.println("");
        sc.close();
    }
}

// TC=O(2*n) SC=O(n)
// TC=O(n) SC=O(1)

File: cpp/Array/zeroestoend.cpp
#include<bits/stdc++.h>
using namespace std;

vector<int> movezeroes(vector<int>a,int n)
{
    // vector<int>temp;
    // for(int i=0;i<n;i++)
    // {
    //     if(a[i]!=0)
    //         temp.push_back(a[i]);
    // }
    // int s=temp.size();
    // for(int i=0;i<s;i++)
    // {
    //     a[i]=temp[i];
    // }
    // for(int i=s;i<n;i++)
    // {
    //     a[i]=0;
    // }
    // return a;

    int j=0;
    for(int i=0;i<n;i++)
    {
        if(a[i]!=0)
        {
            swap(a[i],a[j]);
            j++;
        }
    }
    return a;
}


int main()
{
    int n;
    cin>>n;
    vector<int>a(n);
    for(int i=0;i<n;i++)
    {
        cin>>a[i];
    }
    vector<int>ans = movezeroes(a,n);
    for(auto it:ans)
    {
        cout<<it<<" ";
    }
    return 0;
}

// TC=O(2*n) SC=O(n)
// TC=O(n) SC=O(1)
